### PR TITLE
Add namespace creation and keycloak mangement to tenant controller

### PR DESCRIPTION
**Pull request by GabriFila**
# Description

This PR includes: 
- refactored tenant CRD to rename/remove certain fields
- added namespace creation to tenant controller
- added ketycloak features to tenant controller
  - user creation/update
  - role assignment 

_IMPORTANT NOTE:_ 
the current implementation for the keycloak features assumes that the workspaces of a tenant already exists in the cluster and are OK. In the next PRs I will add the checks in the tenant controller to avoid creating a resource with workspaces that do not exists. I will also add a way to update a tenant when one of his workspaces is deleted.

# How Has This Been Tested?

This PR includes testing for the funcionalities added. The tests related to the keycloak part can be improved, e.g. by covering all the possible cases. At the moment only one case is tested.

### CODE REVIEW

File: operators/pkg/tenant-controller/keycloak_utils.go [hunk 1]
```diff
@@ -3,30 +3,36 @@ package tenant_controller
...
+	Client                gocloak.GoCloak
+	Token                 *gocloak.JWT
+	TargetRealm           string
+	TargetClientID        string
+	UserRequiredActions   []string
+	EmailActionsLifeSpanS int
 }
 
-func createKcRoles(ctx context.Context, kcA *KcActor, rolesToCreate []string) error {
+// temp to add a new commit
```
Comment (palexster) [inline operators/pkg/tenant-controller/keycloak_utils.go:23]: forgotten line?

File: operators/pkg/tenant-controller/keycloak_utils.go [hunk 2]
```diff
@@ -65,3 +71,147 @@ func deleteKcRoles(ctx context.Context, kcA *KcActor, rolesToDelete []string) er
...
+		UserID:   &newUserID,
+		Lifespan: &kcA.EmailActionsLifeSpanS,
+		Actions:  &kcA.UserRequiredActions,
+	}); err != nil {
+		klog.Errorf("Error when sending email actions for user %s", username)
+		klog.Error(err)
+		return nil, err
+	}
+
+	klog.Infof("Sent verification email to user %s", username)
```
Comment (palexster) [inline operators/pkg/tenant-controller/keycloak_utils.go:122]: Do you plan to add some feedback in the Tenant resource status?

**Review by palexster — APPROVED** [review]

File: operators/pkg/tenant-controller/keycloak_utils.go (hunk 1 above)
Comment (GabriFila) [reply operators/pkg/tenant-controller/keycloak_utils.go:23]: Yup, thanks 😅

File: operators/pkg/tenant-controller/keycloak_utils.go (hunk 2 above)
Comment (GabriFila) [reply operators/pkg/tenant-controller/keycloak_utils.go:122]: At the moment no, we'll need to think how to handle the resending of a verification email if the user forgots to confim before the timeout

**Review by giorio94 — APPROVED** [review]
/lgtm

Just a couple of very minor comments

File: operators/api/v1alpha1/tenant_types.go [hunk 3]
```diff
@@ -24,33 +24,31 @@ import (
...
+	// User allows to interact with owned vms
+	User WorkspaceUserRole = "user"
 )
 
 // UserWorkspaceData contains the info of the workspaces related to a user
 type UserWorkspaceData struct {
-	WorkspaceURL string            `json:"workspaceURL"`
-	GroupNumber  int               `json:"groupNumber"`
-	Role         WorkspaceUserRole `json:"role"`
+	Name        string            `json:"name"`
```
[the remark this answers is not in the document]
Comment (palexster) [reply operators/api/v1alpha1/tenant_types.go:39]: Nice catch. Totally agree!

**Pull request merged**